### PR TITLE
feat(tags): Pass environment_id down to GroupSerializer to get proper…

### PR DESCRIPTION
**Pull request by bretthoerner**
… user counts

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by bretthoerner** [comment]
@tkaemming when you have 💉 ☕️  today I have some observations.

* I have to pass `environment_id` down to `GroupSerializer` which is used in *many* places (see diff).
* This forces me to manually import/pass the serializer class so I can pass the argument down.
* I had to move the `environment_id_func` bit up to `Serializer` because there are serializers that themselves call `GroupSerializer` (e.g. `OrganizationActivitySerializer`)
* I'm starting to feel like `EnvironmentMixin` is being used everywhere. And I'm only/solely concerned with tags so far.

So I guess I'm starting to wonder:

* Whether the `EnvironmentMixin` shouldn't be available automatically in more/all endpoints. When people start caring out things other than TSDB and tags I feel like it will be pervasive. I'm mostly indifferent on it being explicit like it is now, though.
* Whether `environment_id_func` should just be passed as a kwarg to `serialize` so we don't have to import/use the serializer classes manually.
* What my purpose on 🌍 is.

**Comment by tkaemming** [comment]
> * I have to pass `environment_id` down to `GroupSerializer` which is used in *many* places (see diff). This forces me to manually import/pass the serializer class so I can pass the argument down.

I don't think this is inherently bad. (Personally, I prefer it, since it makes it more obvious what's happening, but that's just a preference and not something that's objectively better or worse.) More on this point below.

> * I had to move the `environment_id_func` bit up to `Serializer` because there are serializers that themselves call `GroupSerializer` (e.g. `OrganizationActivitySerializer`)

I'm torn on this as a design decision rather than adding it as a constructor parameter to only the classes that utilize it, since I don't think every subclass of `Serializer` actually requires (or will require) the environment function, and unused instance state is/can be confusing. At the same time, I understand the pragmatic perspective that this is a lot simpler than adding it only specifically to things that require it. This seems at odds with the decision above to explicitly pass the serializer instance to the `serialize` calls though.

My thinking would be to either:

1. Remove the parameter from `Serializer` and only include it on the classes that need it (including the ones that need it transitively) and continue passing the serializer instance as a parameter to `serialize`, or,
2. Just blindly [pass it through as a keyword argument](https://github.com/getsentry/sentry/blob/345ff81ecd853bbf9d2ff9f3cba37f6bd08350ea/src/sentry/api/serializers/base.py#L8) and make sure the things that need it can accept it there, or at least don't fail if they don't need it. (Personally, this option makes me cringe but I can see why it could be useful.)

> * I'm starting to feel like `EnvironmentMixin` is being used everywhere. And I'm only/solely concerned with tags so far.

When you say "concerned with tags so far" is there something else you're anticipating being concerned about? 
 
> * Whether the `EnvironmentMixin` shouldn't be available automatically in more/all endpoints. When people start caring out things other than TSDB and tags I feel like it will be pervasive. I'm mostly indifferent on it being explicit like it is now, though.

I basically see the same pattern here as with the serializers. It's definitely "simpler" (when initially writing the code, at least) to just include the ability to grab an environment out of any API that is scoped to at least an organization. Personally, I don't think that outweighs the weirdness of adding that to API endpoints that don't need it (e.g. the ingestion statistic endpoints, other management stuff, etc.) There's probably a tipping point somewhere, but I'm not sure where that is (and again, is totally subjective.)

> * Whether `environment_id_func` should just be passed as a kwarg to `serialize` so we don't have to import/use the serializer classes manually. Or if there's another way to reduce the duplication/explicitness going on. It feels like almost every serializer may need the `environment_id` soon.

I think I answered this one. 🙂 

> * What my purpose on 🌍 is.

**Comment by tkaemming** [comment]
I guess the tl;dr of the above with my perspective (since I just realized I said a lot of words that provided some context around my thinking but didn't actually articulate any viewpoints): If it were my choice, I'd keep passing around `environment_id_func` through serializer constructors as directly as possible (and even more directly than you were talking about here, see the comments about `Serializer` above) since I'm not convinced yet that the API usage of the environment is as widespread as we think: I think we just feel that way since we've been knee deep in those endpoints and none (or comparably few) of the other ones for a few weeks.

**Comment by bretthoerner** [comment]
> When you say "concerned with tags so far" is there something else you're anticipating being concerned about?

No, I mean concerned as in I have solely been thinking about tags. I feel like more things will need to be scoped by environment but I haven't given any of them the time of day just yet.

---

I agree with the other bits. I was worried making a bunch of things that were implicit into explicit was a bad style move (given how the code looked before). I'm all for more explicit in the end, though.

File: src/sentry/api/endpoints/issues_resolved_in_release.py [hunk 1]
```diff
@@ -61,5 +61,13 @@ def get(self, request, project, version):
 
         groups = Group.objects.filter(project=project, id__in=group_ids)
 
-        context = serialize(list(groups), request.user, StreamGroupSerializer(stats_period=None))
+        context = serialize(
+            list(groups),
+            request.user,
+            StreamGroupSerializer(
+                stats_period=None,
```
Comment (tkaemming) [inline src/sentry/api/endpoints/issues_resolved_in_release.py]: These aren't really specific to this change, but:

- `stats_period` defaults to `None` and doesn't need to be provided here,
- the behavior of `GroupSerializer` and `StreamGroupSerializer` is only differentiated when `stats_period` or `matching_event_id` is not `None`. Neither of those are the case here, so shouldn't this use `GroupSerializer`?

We should go through and clean these up later.

File: src/sentry/api/endpoints/organization_activity.py [hunk 2]
```diff
@@ -27,5 +28,8 @@ def get(self, request, organization, member):
             queryset=queryset,
             paginator_cls=DateTimePaginator,
             order_by='-datetime',
-            on_results=lambda x: serialize(x, request.user, OrganizationActivitySerializer()),
+            on_results=lambda x: serialize(x, request.user, OrganizationActivitySerializer(
```
Comment (tkaemming) [inline src/sentry/api/endpoints/organization_activity.py]: Lovely formatting, this

File: src/sentry/api/endpoints/team_groups_new.py [hunk 3]
```diff
@@ -41,4 +42,17 @@ def get(self, request, team):
         for group in group_list:
             group._project_cache = project_dict.get(group.project_id)
 
-        return Response(serialize(group_list, request.user))
+        if group_list:
+            organization_id = group_list[0].project.organization_id
```
Comment (tkaemming) [inline src/sentry/api/endpoints/team_groups_new.py]: Isn't it safer to get this off of `team`?

File: src/sentry/api/endpoints/team_groups_trending.py [hunk 4]
```diff
@@ -41,4 +42,18 @@ def get(self, request, team):
         for group in group_list:
             group._project_cache = project_dict.get(group.project_id)
 
-        return Response(serialize(group_list, request.user))
+        if group_list:
+            organization_id = group_list[0].project.organization_id
```
Comment (tkaemming) [inline src/sentry/api/endpoints/team_groups_trending.py]: Same comment as above with regard to `team`.

File: src/sentry/api/serializers/models/activity.py [hunk 5]
```diff
@@ -10,6 +10,9 @@
 
 @register(Activity)
 class ActivitySerializer(Serializer):
+    def __init__(self, environment_id_func=None):
+        self.environment_id_func = environment_id_func if environment_id_func is not None else lambda: None
```
Comment (tkaemming) [inline src/sentry/api/serializers/models/activity.py]: This should be `None` and allow the `GroupSerializer` constructor to manage the default value.

File: src/sentry/api/serializers/models/userreport.py [hunk 6]
```diff
@@ -59,10 +59,18 @@ def serialize(self, obj, attrs, user):
 
 
 class ProjectUserReportSerializer(UserReportSerializer):
+    def __init__(self, environment_id_func=None):
+        self.environment_id_func = environment_id_func if environment_id_func is not None else lambda: None
```
Comment (tkaemming) [inline src/sentry/api/serializers/models/userreport.py]: This should be `None` and allow the `GroupSerializer` constructor to manage the default value.

File: src/sentry/models/group.py [hunk 7]
```diff
@@ -352,22 +352,6 @@ def get_oldest_event(self):
                 self._oldest_event = None
         return self._oldest_event
 
-    def get_tags(self):
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply src/sentry/models/group.py]: :nice:

**Review by tkaemming — CHANGES_REQUESTED** [review]

**Comment by bretthoerner** [comment]
All good finds. ⚔️

File: src/sentry/api/endpoints/issues_resolved_in_release.py (hunk 1 above)
Comment (tkaemming) [reply src/sentry/api/endpoints/issues_resolved_in_release.py]: Or here too this works 😎

**Review by tkaemming — APPROVED** [review]
Punch it 🎸

File: src/sentry/api/endpoints/issues_resolved_in_release.py (hunk 1 above)
Comment (bretthoerner) [reply src/sentry/api/endpoints/issues_resolved_in_release.py]: I like to keep my contribution graph nice and low.

**Pull request merged**